Fix: 로그아웃 기능 수정 및 리프레시 토큰

Hujin0322 opened the pull request:
## #️⃣ Issue Number

- #50

## 📝 요약(Summary)

- 로그인: 액세스 토큰(30분)과 리프레시 토큰(1일) 발급, 리프레시 토큰은 쿠키에 저장
- 토큰 재발급: 만료 전 액세스 토큰 블랙리스트 추가, 새로운 토큰 발급 후 기존 리프레시 토큰 블랙리스트 추가
- 로그아웃: 리프레시 토큰 블랙리스트 추가, 쿠키 삭제
- 게이트웨이 인증 필터: 블랙리스트 확인 후 유효한 토큰의 정보 추출 및 요청 전달

❗️블랙리스트에 토큰을 추가할 때, 해당 토큰의 남은 유효기간을 확인하고, 그 기간 동안만 블랙리스트에 유지되도록 설정했습니다.

## 💬 공유사항 to 리뷰어

- 나름 고민하여 추가해봤는데, 확신이 서지 않아 의견을 여쭙고 싶습니다. 
혹시 부족하거나 이상한 부분이 있다면 편하게 말씀 부탁드립니다!

## 📷 스크린샷

<!--- 선택사항 -->

- 토큰 재발급: 리프레시 토큰 쿠키 변경 확인, 액세스 토큰 발급 확인, 블랙 리스트에 추가됨 확인 
![스크린샷 2025-01-16 오전 11 06 14](https://github.com/user-attachments/assets/4cddf2b2-a41c-489e-8d9d-d98c8c543497)

- 로그아웃: 리프레시 토큰 쿠키 삭제됨 확인, 블랙리스트 추가됨 확인
![스크린샷 2025-01-16 오전 11 07 44](https://github.com/user-attachments/assets/87381c45-b95a-48bc-9a6f-4c6c733016db)


## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Hujin0322 committed:
Merge pull request #49 from Cool-Potatoes/feat/user/manage-password-id

Feat: 아이디, 비밀번호 관련 기능 구현

Hujin0322 committed:
refactor: 인증 코드 생성 로직을 private 메서드로 변경

Hujin0322 committed:
Merge remote-tracking branch 'origin/dev' into dev

Hujin0322 committed:
Merge remote-tracking branch 'origin/dev' into dev

Hujin0322 committed:
Merge branch 'dev' into fix/user/logout

Hujin0322 committed:
chore: 불필요한 어노테이션 및 파일 제거

Hujin0322 committed:
refactor: 인증 필요없는 경로 리스트 수정

Hujin0322 committed:
refactor: userKafkaSender 적용

Hujin0322 committed:
Merge remote-tracking branch 'origin/dev' into dev

Hujin0322 committed:
Merge branch 'refs/heads/dev' into fix/user/logout

Hujin0322 committed:
Merge remote-tracking branch 'origin/dev' into dev

Hujin0322 committed:
Merge branch 'refs/heads/dev' into fix/user/logout

Hujin0322 committed:
refactor: 비밀번호 재설정 응답 문구 수정

Hujin0322 committed:
refactor: 토큰 별 만료기한 변경

Hujin0322 committed:
refactor: 카프카 이벤트 발행 시 메소드명과 클래스명 추가

Hujin0322 committed:
refactor: 토큰 재발급 인증 제외 경로에 추가

Hujin0322 committed:
refactor: 로그인 기능 수정

- RefreshToken 생성 후 쿠키에 저장

Hujin0322 committed:
feat: 토큰 관련 에러 코드 추가

Hujin0322 committed:
feat: 토큰 재발급

- 기존 Access Token이 만료되지 않은 경우 블랙리스트에 추가
- 기존 Refresh Token 블랙리스트에 추가
- 새 토큰 생성 및 Refresh Token 쿠키에 추가

Hujin0322 committed:
refactor: 로그아웃 기능 수정

- 블랙리스트 존재하는 토큰인지 확인 후 로그아웃
- 쿠키 삭제

Hujin0322 committed:
feat: gateway에 Redis 설정 추가

Hujin0322 committed:
feat: 에러 코드 추가

Hujin0322 committed:
feat: gateway 인증필터에 블랙리스트 확인 로직 추가

Hujin0322 committed:
Merge remote-tracking branch 'origin/dev' into dev

Hujin0322 committed:
Merge remote-tracking branch 'origin/dev' into dev

k-jaehyun commented on user-service/src/main/java/com/flight_booking/user_service/application/service/VerificationCodeService.java:
```diff
@@ -57,7 +57,7 @@ public void sendVerificationCode(SendCodeRequest request) {
-    kafkaTemplate.send(
+    userKafkaSender.sendMessage(
         "password-reset-topic", email,
         new NotificationRequestDto(userId, email, code));
```
유저쪽 적용해주셨는데 알림쪽 적용이 보이지 않아 코멘트 남겨놓습니다!

현재 kafkaSender 구조가 바뀌었는데 반영이 되었을까요?

Hujin0322 replied:
user쪽만 신경쓰고 알림쪽을 놓쳤네요!!!
반영해서 추가 커밋하겠습니다!

k-jaehyun commented on user-service/src/main/java/com/flight_booking/user_service/application/service/AuthService.java:
```diff
@@ -57,40 +52,56 @@ public void createUser(SignUpRequest request) {
+
+      // 인증 처리
       Authentication authentication = authenticationManager.authenticate(
           new UsernamePasswordAuthenticationToken(email, password)
       );
-      CustomUserDetails userDetails = (CustomUserDetails) authentication.getPrincipal();
 
-      String validatedEmail = userDetails.getUsername();
+      CustomUserDetails userDetails = (CustomUserDetails) authentication.getPrincipal();
       String role = userDetails.getUser().getRole().toString();
```
로그인 과정에서 인증객체 생성 및 주입은 필요하지 않은 것 같습니다.
어떤 의도로 생성 및 주입하신걸까요??
Role이 필요하다면 조회한 user에서 가져올 수 있구요!

Hujin0322 replied:
말씀해주신 부분 수정 완료했습니다!

k-jaehyun commented on user-service/src/main/java/com/flight_booking/user_service/infrastructure/security/jwt/JwtUtil.java:
```diff
@@ -42,17 +45,30 @@ public void init() {
+        .issuedAt(now) // 발급 시간
+        .expiration(expirationDate) // 만료 시간
         .signWith(key, SIG.HS256)
         .compact();
```
JWT임을 나타내는 BEARER_PREFIX 가 빠진 것 같습니다!

Hujin0322 replied:
리프레시 토큰은 바로 쿠키에 저장시킬거라 BEARER_PREFIX를 뺐습니다.

k-jaehyun replied:
쿠키에 저장 할 때도 토큰임을 나타내기 위해 보통 BEARER_PREFIX를 넣어줍니다!

Hujin0322 replied:
BEARER_PREFIX = "Bearer "로 추가하는 경우, 띄어쓰기로 인해 발생하는 오류를 방지하기 위해 제외하게 되었습니다.
또한, 검색과 튜터님께 문의한 결과, BEARER_PREFIX는 주로 Authorization 헤더에 액세스 토큰을 전달할 때 사용되며, 리프레시 토큰은 보통 쿠키에 저장되거나 클라이언트에서 서버로 직접 전달되는 방식으로 활용되기 때문에, 이 경우에는 BEARER_PREFIX를 생략하는 것이 일반적이라는 의견을 들었습니다.

이러한 이유로 리프레시 토큰에서는 BEARER_PREFIX를 제외하는 것이 더 적합하다고 생각되는데 혹시 제외하는 방향은 어떠실까요?

k-jaehyun replied:
오 그렇군요!! 죄송합니다 제가 잘못 알고 있었던것 같습니다..!
저렇게 확신에 차서 말하다니 부끄럽네요... 반성하겠습니다😥

Hujin0322 replied:
아니에요!! 사실 오류 때문에 튜터님께 여쭤봤었는데 저런 답변을 받은거라 저도 얻어걸렸습니다 ㅎㅎ
덕분에 좋은 정보 하나 얻어갑니다!!

k-jaehyun commented on user-service/src/main/java/com/flight_booking/user_service/application/service/AuthService.java:
```diff
@@ -105,29 +116,15 @@ private void checkUserStatus(String email) {
-      log.info("토큰 존재함: {}", token);
-      redisTemplate.delete(token);
-    }
+  // RefreshToken을 쿠키에 저장
+  private void addRefreshTokenToCookie(String refreshToken, HttpServletResponse response) {
+    log.info("쿠키 설정 값: {}", refreshToken);
+    Cookie cookie = new Cookie("refreshToken", refreshToken);
+    cookie.setHttpOnly(true);   // 클라이언트에서 접근 불가
+//      cookie.setSecure(true);     // HTTPS에서만 전송 (현재 HTTP)
+    cookie.setPath("/");        // 쿠키 경로
+    cookie.setMaxAge(86400);    // 만료 시간 (1일)
+    response.addCookie(cookie);
+    log.info("Refresh token 쿠키가 성공적으로 설정되었습니다.");
```
`cookie.setHttpOnly(true);` 로 보안을 강화하는 방법도 있군요!!👍👍

k-jaehyun replied:
해당 메서드는 jwtUtil에 있으면 더 깔끔할 것 같습니다!

k-jaehyun replied:
refresh token의 유효시간을 설정파일에서 넣어주셨는데 이곳에는 하드코딩 되어있네요~

k-jaehyun replied:
쿠키 이름은 보통 상수로 선언해서 발행합니다!
`AUTHORIZATION_HEADER = "Authorization"` 라고 해주셨던 것 처럼요

Hujin0322 replied:
확인 후 수정완료했습니다!

k-jaehyun commented on user-service/src/main/java/com/flight_booking/user_service/presentation/controller/AuthController.java:
```diff
@@ -44,4 +46,15 @@ public ApiResponse<?> findId(@Valid @RequestBody FindIdRequest request) {
+
+    String newAccessToken = authService.refreshAccessToken(refreshToken, accessToken, response);
+    return ApiResponse.ok(newAccessToken, "AccessToken 발급 성공");
+  }
```
게이트웨이 통해서 테스트 해보셨을까요??

Hujin0322 replied:
넵!

k-jaehyun commented on user-service/src/main/java/com/flight_booking/user_service/application/service/AuthService.java:
```diff
@@ -127,4 +177,22 @@ private void addRefreshTokenToCookie(String refreshToken, HttpServletResponse re
+      log.warn("만료된 토큰입니다. 토큰: {}", token);
+      throw new UserException(ErrorCode.TOKEN_EXPIRED);
+    }
+  }
```
마찬가지로 jwt 관련된 로직이므로 jwtUtil에 있었으면 하는 아쉬움이 있습니다~

Hujin0322 replied:
이동 완료했습니다!

k-jaehyun commented on user-service/src/main/java/com/flight_booking/user_service/application/service/AuthService.java:
```diff
@@ -97,6 +103,50 @@ public String findId(FindIdRequest request) {
+    UserDetails userDetails = customUserDetailsService.loadUserByUsername(email);
+    String role = userDetails.getAuthorities().toString();
+
+    // 새로운 Access Token, Refresh Token 발급
+    String newAccessToken = jwtUtil.createAccessToken(email, role);
+    String newRefreshToken = jwtUtil.createRefreshToken(email);
```
메서드 이름이 refreshAccessToken 여서 access만 리프레시 하는줄알았는데,
모두 함께 리프레시 되는군요!

Hujin0322 replied:
메서드명이 혼란을 주는 거 같아 renewTokens로 변경했습니다!

k-jaehyun commented on user-service/src/main/java/com/flight_booking/user_service/application/service/AuthService.java:
```diff
@@ -147,6 +147,23 @@ public String refreshAccessToken(String refreshToken, String accessToken,
     return newAccessToken;
   }
 
+  // 로그아웃
+  public void logout(String token, HttpServletResponse response) {
+    if (isTokenBlacklisted(token)) {
+      throw new UserException(ErrorCode.BLACKLISTED_TOKEN);
+    }
+
+    addToBlacklist(token);
+
+    // 쿠키 삭제
+    Cookie cookie = new Cookie("refreshToken", null);
+    cookie.setHttpOnly(true);
+    cookie.setSecure(false);
+    cookie.setPath("/");
+    cookie.setMaxAge(0);  // 쿠키 만료
+    response.addCookie(cookie);
+  }
+
```
Access Token은 여전히 살아있게 될 것 같습니다!

Hujin0322 replied:
헤더에 있는 액세스 토큰과 쿠키의 리프레시 토큰 둘 다 가져와
블랙리스트에 존재하는지 확인 후 블랙리스트에 추가하고 쿠키 삭제하는 것으로 수정했습니다.

k-jaehyun commented on gateway-service/src/main/java/com/flight_booking/gateway_service/filter/JwtAuthenticationFilter.java:
```diff
@@ -40,25 +40,36 @@ public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
-    String email = jwtUtil.extractEmail(token);
-    String role = jwtUtil.extractRole(token);
+    try {
+      jwtUtil.validateToken(token);
+      log.info("토큰 검증 완료");
```
inValidate한 토큰에 대한 처리가 이뤄지지 않을 것 같습니다!
내부에서 try-catch로 구현되어 있네요

k-jaehyun commented on gateway-service/src/main/java/com/flight_booking/gateway_service/filter/JwtAuthenticationFilter.java:
```diff
@@ -40,25 +40,36 @@ public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
-        .header("X-USER-EMAIL", email)
-        .header("X-USER-ROLE", role)
-        .build();
+      jwtUtil.isTokenBlacklisted(token);
+      log.info("블랙리스트 체크 완료");
```
내부에서 exception을 던져주지만 catch 부분과 다른 것 같습니다!

k-jaehyun reviewed (approved):
로그아웃 및 refresh token을 이렇게 구현하셨군요! 
인증필터에서도 에러를 catch 해주셔서 이제 에러 원인도 잘 알 수 있겠어요~
고생하셨습니다!! 👍👍

Hujin0322 committed:
refactor: 토큰 관련 코드 리팩토링

- 토큰 관련 메서드를 jwtUtil로 위치 이동
- 로그인: 토큰 발급 흐름 수정
- 토큰 재발급: 메서드명 renewTokens로 변경
- 로그아웃: 액세스 토큰의 블랙리스트 확인 로직 추가

Hujin0322 committed:
refactor: kafka 설정 변경 및 정리

- Kafka 메시지의 데이터 타입을 ApiResponse<?>으로 변경
- NotificationKafkaEndpoint와 NotificationKafkaSender 생성
- 이메일 전송 응답 메세지 추가

Hujin0322 committed:
refactor: 리프레시 토큰 쿠키 이름 변경

Hujin0322 committed:
Merge remote-tracking branch 'origin/dev' into dev

Hujin0322 committed:
Merge branch 'refs/heads/dev' into fix/user/logout

# Conflicts:
#	user-service/src/main/java/com/flight_booking/user_service/infrastructure/configuration/kafka/ProducerApplicationKafkaConfig.java
